File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonBinaryReader.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.io.DsonIOException;
import cn.wjybxx.dson.io.DsonInput;
import cn.wjybxx.dson.types.*;

import java.util.Objects;

/**
 * @author wjybxx
 * date - 2023/4/22
 */
public class DsonBinaryReader extends AbstractDsonReader {

    private DsonInput input;

    public DsonBinaryReader(DsonReaderSettings settings, DsonInput input) {
        super(settings);
        this.input = Objects.requireNonNull(input);
        setContext(new Context().init(null, DsonContextType.TOP_LEVEL, null));
    }

    @Override
    protected Context getContext() {
        return (Context) super.getContext();
    }

    @Override
    protected AbstractDsonReader.Context newContext() {
        return new Context();
    }

    @Override
    public void close() {
        if (setting.autoClose && input != null) {
            input.close();
            input = null;
        }
        super.close();
    }

    // region state

    @Override
    public DsonType readDsonType() {
        Context context = this.getContext();
        checkReadDsonTypeState(context);

        final int fullType = input.isAtEnd() ? 0 : Byte.toUnsignedInt(input.readRawByte());
        final int wreTypeBits = Dsons.wireTypeOfFullType(fullType);
        DsonType dsonType = DsonType.forNumber(Dsons.dsonTypeOfFullType(fullType));
        WireType wireType = dsonType.hasWireType() ? WireType.forNumber(wreTypeBits) : WireType.VARINT;
        this.currentDsonType = dsonType;
        this.currentWireType = wireType;
        this.currentWireTypeBits = wreTypeBits;
        this.currentName = INVALID_NAME;

        onReadDsonType(context, dsonType);
        return dsonType;
    }

    @Override
    public DsonType peekDsonType() {
        Context context = this.getContext();
        checkReadDsonTypeState(context);

        final int fullType = input.isAtEnd() ? 0 : Byte.toUnsignedInt(input.getByte(input.getPosition()));
        return DsonType.forNumber(Dsons.dsonTypeOfFullType(fullType));
    }

    @Override
    protected void doReadName() {
        String fieldName = input.readString();
        if (setting.enableFieldIntern) {
            currentName = Dsons.internField(fieldName);
        } else {
            currentName = fieldName;
        }
    }

    // endregion

    // region 简单值

    @Override
    protected int doReadInt32() {
        return currentWireType.readInt32(input);
    }

    @Override
    protected long doReadInt64() {
        return currentWireType.readInt64(input);
    }

    @Override
    protected float doReadFloat() {
        return DsonReaderUtils.readFloat(input, currentWireTypeBits);
    }

    @Override
    protected double doReadDouble() {
        return DsonReaderUtils.readDouble(input, currentWireTypeBits);
    }

    @Override
    protected boolean doReadBool() {
        return DsonReaderUtils.readBool(input, currentWireTypeBits);
    }

    @Override
    protected String doReadString() {
        return input.readString();
    }

    @Override
    protected void doReadNull() {

    }

    @Override
    protected Binary doReadBinary() {
        return DsonReaderUtils.readBinary(input);
    }

    @Override
    protected ExtInt32 doReadExtInt32() {
        return DsonReaderUtils.readExtInt32(input, currentWireType);
    }

    @Override
    protected ExtInt64 doReadExtInt64() {
        return DsonReaderUtils.readExtInt64(input, currentWireType);
    }

    @Override
    protected ExtDouble doReadExtDouble() {
        return DsonReaderUtils.readExtDouble(input, currentWireTypeBits);
    }

    @Override
    protected ExtString doReadExtString() {
        return DsonReaderUtils.readExtString(input, currentWireTypeBits);
    }

    @Override
    protected ObjectRef doReadRef() {
        return DsonReaderUtils.readRef(input, currentWireTypeBits);
    }

    @Override
    protected OffsetTimestamp doReadTimestamp() {
        return DsonReaderUtils.readTimestamp(input);
    }

    // endregion

    // region 容器

    @Override
    protected void doReadStartContainer(DsonContextType contextType, DsonType dsonType) {
        Context newContext = newContext(getContext(), contextType, dsonType);
        int length = input.readFixed32();
        newContext.oldLimit = input.pushLimit(length);
        newContext.name = currentName;

        this.recursionDepth++;
        setContext(newContext);
    }

    @Override
    protected void doReadEndContainer() {
        if (!input.isAtEnd()) {
            throw DsonIOException.bytesRemain(input.getBytesUntilLimit());
        }
        Context context = getContext();
        input.popLimit(context.oldLimit);

        // 恢复上下文
        recoverDsonType(context);
        this.recursionDepth--;
        setContext(context.parent);
        returnContext(context);
    }

    // endregion

    // region 特殊接口

    @Override
    protected void doSkipName() {
        // 避免构建字符串
        int size = input.readUint32();
        if (size > 0) {
            input.skipRawBytes(size);
        }
    }

    @Override
    protected void doSkipValue() {
        DsonReaderUtils.skipValue(input, getContextType(), currentDsonType, currentWireType, currentWireTypeBits);
    }

    @Override
    protected void doSkipToEndOfObject() {
        DsonReaderUtils.skipToEndOfObject(input);
    }

    @Override
    protected byte[] doReadValueAsBytes() {
        return DsonReaderUtils.readValueAsBytes(input, currentDsonType);
    }

    // endregion

    // region context

    private Context newContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context context = (Context) rentContext();
        context.init(parent, contextType, dsonType);
        return context;
    }

    protected static class Context extends AbstractDsonReader.Context {

        int oldLimit = -1;

        public Context() {
        }

        public void reset() {
            super.reset();
            oldLimit = -1;
        }
    }

    // endregion

}

File: csharp/Dson/src/DsonBinaryReader.cs
#region LICENSE

//  Copyright 2023 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using Wjybxx.Dson.IO;
using Wjybxx.Dson.Types;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

/// <summary>
/// Dson二进制Reader
/// </summary>
/// <typeparam name="TName"></typeparam>
public class DsonBinaryReader<TName> : AbstractDsonReader<TName> where TName : IEquatable<TName>
{
#nullable disable
    private IDsonInput _input;
#nullable enable
    private readonly AbstractDsonReader<string>? _textReader;
    private readonly AbstractDsonReader<FieldNumber>? _binReader;

    public DsonBinaryReader(DsonReaderSettings settings, IDsonInput input) : base(settings) {
        this._input = input ?? throw new ArgumentNullException(nameof(input));
        SetContext(new Context().Init(null, DsonContextType.TopLevel, DsonTypes.Invalid));
        if (DsonInternals.IsStringKey<TName>()) {
            this._textReader = this as AbstractDsonReader<string>;
            this._binReader = null;
        } else {
            this._textReader = null;
            this._binReader = this as AbstractDsonReader<FieldNumber>;
        }
    }

    protected override Context GetContext() {
        return (Context)_context;
    }

    protected override Context? GetPooledContext() {
        return (Context?)base.GetPooledContext();
    }

    public override void Dispose() {
        if (Settings.AutoClose) {
            _input?.Dispose();
            _input = null!;
        }
        base.Dispose();
    }

    #region state

    public override DsonType ReadDsonType() {
        Context context = GetContext();
        CheckReadDsonTypeState(context);

        int fullType = _input.IsAtEnd() ? 0 : BinaryUtils.ToUint(_input.ReadRawByte());
        int wreTypeBits = Dsons.WireTypeOfFullType(fullType);
        DsonType dsonType = DsonTypes.ForNumber(Dsons.DsonTypeOfFullType(fullType));
        WireType wireType = dsonType.HasWireType() ? WireTypes.ForNumber(wreTypeBits) : WireType.VarInt;
        this._currentDsonType = dsonType;
        this._currentWireType = wireType;
        this._currentWireTypeBits = wreTypeBits;
        this._currentName = default!;

        OnReadDsonType(context, dsonType);
        return dsonType;
    }

    public override DsonType PeekDsonType() {
        Context context = GetContext();
        CheckReadDsonTypeState(context);

        int fullType = _input.IsAtEnd() ? 0 : BinaryUtils.ToUint(_input.GetByte(_input.Position));
        return DsonTypes.ForNumber(Dsons.DsonTypeOfFullType(fullType));
    }

    protected override void DoReadName() {
        if (_textReader != null) {
            string filedName = _input.ReadString();
            if (Settings.EnableFieldIntern) {
                filedName = Dsons.InternField(filedName);
            }
            _textReader._currentName = filedName;
        } else {
            _binReader!._currentName = FieldNumber.OfFullNumber(_input.ReadUint32());
        }
    }

    #endregion

    #region 简单值

    protected override int DoReadInt32() {
        return DsonReaderUtils.ReadInt32(_input, _currentWireType);
    }

    protected override long DoReadInt64() {
        return DsonReaderUtils.ReadInt64(_input, _currentWireType);
    }

    protected override float DoReadFloat() {
        return DsonReaderUtils.ReadFloat(_input, _currentWireTypeBits);
    }

    protected override double DoReadDouble() {
        return DsonReaderUtils.ReadDouble(_input, _currentWireTypeBits);
    }

    protected override bool DoReadBool() {
        return DsonReaderUtils.ReadBool(_input, _currentWireTypeBits);
    }

    protected override string DoReadString() {
        return _input.ReadString();
    }

    protected override void DoReadNull() {
    }

    protected override Binary DoReadBinary() {
        return DsonReaderUtils.ReadBinary(_input);
    }

    protected override ExtInt32 DoReadExtInt32() {
        return DsonReaderUtils.ReadExtInt32(_input, _currentWireType);
    }

    protected override ExtInt64 DoReadExtInt64() {
        return DsonReaderUtils.ReadExtInt64(_input, _currentWireType);
    }

    protected override ExtDouble DoReadExtDouble() {
        return DsonReaderUtils.ReadExtDouble(_input, _currentWireTypeBits);
    }

    protected override ExtString DoReadExtString() {
        return DsonReaderUtils.ReadExtString(_input, _currentWireTypeBits);
    }

    protected override ObjectRef DoReadRef() {
        return DsonReaderUtils.ReadRef(_input, _currentWireTypeBits);
    }

    protected override OffsetTimestamp DoReadTimestamp() {
        return DsonReaderUtils.ReadTimestamp(_input);
    }

    #endregion

    #region 容器

    protected override void DoReadStartContainer(DsonContextType contextType, DsonType dsonType) {
        Context newContext = NewContext(GetContext(), contextType, dsonType);
        int length = _input.ReadFixed32();
        newContext.oldLimit = _input.PushLimit(length);
        newContext.name = _currentName;

        this._recursionDepth++;
        SetContext(newContext);
    }

    protected override void DoReadEndContainer() {
        if (!_input.IsAtEnd()) {
            throw DsonIOException.BytesRemain(_input.GetBytesUntilLimit());
        }
        Context context = GetContext();
        _input.PopLimit(context.oldLimit);

        // 恢复上下文
        RecoverDsonType(context);
        this._recursionDepth--;
        SetContext(context.parent!);
        PoolContext(context);
    }

    #endregion

    #region 特殊

    protected override void DoSkipName() {
        if (_textReader != null) {
            // 避免构建字符串
            int size = _input.ReadUint32();
            if (size > 0) {
                _input.SkipRawBytes(size);
            }
        } else {
            _input.ReadUint32();
        }
    }

    protected override void DoSkipValue() {
        DsonReaderUtils.SkipValue(_input, ContextType, _currentDsonType, _currentWireType, _currentWireTypeBits);
    }

    protected override void DoSkipToEndOfObject() {
        DsonReaderUtils.SkipToEndOfObject(_input);
    }

    protected override byte[] DoReadValueAsBytes() {
        return DsonReaderUtils.ReadValueAsBytes(_input, _currentDsonType);
    }

    #endregion

    #region context

    private Context NewContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context? context = GetPooledContext();
        if (context != null) {
            SetPooledContext(null);
        } else {
            context = new Context();
        }
        context.Init(parent, contextType, dsonType);
        return context;
    }

    private void PoolContext(Context context) {
        context.Reset();
        SetPooledContext(context);
    }

    protected new class Context : AbstractDsonReader<TName>.Context
    {
        protected internal int oldLimit = -1;

        public Context() {
        }

        public override void Reset() {
            base.Reset();
            oldLimit = -1;
        }
    }

    #endregion
}